### PR TITLE
Support bf16 ScatterOp

### PR DESCRIPTION
**Pull request by eedalong**

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/mhlo_decomp_rewriters.cc [hunk 1]
```diff
@@ -132,6 +132,104 @@ LogicalResult SliceOpConvert::matchAndRewrite(mhlo::SliceOp op,
...
+
+    auto scatter_op = rewriter.create<mhlo::ScatterOp>(
+        hloOp.getLoc(), self_fp32_type, new_operands, hloOp->getAttrs());
+
+    auto res_convert_back = rewriter.create<mhlo::ConvertOp>(
+        hloOp.getLoc(), self_bf16_type, scatter_op->getResult(0));
+
+    // Construct updateComputation region, here we treat it as update operation
+    Block& block = scatter_op.getUpdateComputation().emplaceBlock();
+    auto blockArg1Type =
```
Comment (Yancey1989) [inline tao_compiler/mlir/disc/transforms/mhlo_decomp_rewriters.cc:189]: arg1 and arg2 have the same type, we don't need two type var.

File: tao_compiler/mlir/disc/transforms/mhlo_decomp_rewriters.cc [hunk 2]
```diff
@@ -132,6 +132,104 @@ LogicalResult SliceOpConvert::matchAndRewrite(mhlo::SliceOp op,
...
+    auto blockArg2Type =
+        RankedTensorType::get({}, src_fp32_type.getElementType());
+    block.addArgument(blockArg1Type, hloOp.getLoc());
+    block.addArgument(blockArg2Type, hloOp.getLoc());
+    {
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.setInsertionPointToStart(&block);
+
+      Value retValue;
+      if (llvm::isa<mhlo::AddOp>(calc_op)) {
```
Comment (Yancey1989) [inline tao_compiler/mlir/disc/transforms/mhlo_decomp_rewriters.cc:200]: can we polish this code block using a lambda function `createOp`?
``` cpp
auto createOp = [&](auto opType) {
    return rewriter.create<decltype(opType)>(op->getLoc(), block.getArgument(0), block.getArgument(1)).getResult();
};

if (llvm::isa<mhlo::AddOp>(calc_op)) {
    retValue = createOp(mhlo::AddOp());
}
```

File: tao_compiler/mlir/disc/transforms/mhlo_decomp_rewriters.cc (hunk 1 above)
Comment (eedalong) [reply tao_compiler/mlir/disc/transforms/mhlo_decomp_rewriters.cc]: Fixed.

File: tao_compiler/mlir/disc/transforms/mhlo_decomp_rewriters.cc (hunk 2 above)
Comment (eedalong) [reply tao_compiler/mlir/disc/transforms/mhlo_decomp_rewriters.cc]: Fixed.

**Review by Yancey1989 — APPROVED** [review]
LGTM

**Pull request merged**